Stdenv setup: Process dependencies before dependents

Ericson2314 opened the pull request:
###### Motivation for this change

This is intended to be reviewed commit-by-commit.

This does 2 things

1. Change the dependency order. Before, we did the current node before children, now we do the children before current node. Were dependencies forming a tree, this would simply be the inverse traversal, but since they in fact form a DAG, it's a bit more subtle. The opposite of the new invariant "dependencies before dependents" wasn't true: dependents weren't always processed first because it was possible that a dependency was already visited via another dependent.

   This is an observable change because the processing we are doing does not commute: arbitrary bash setup hooks certainly don't, and `PATH` concatenation doesn't either. I do think having this invariant is valuable, though, and rely on it to simplify cc-wrapper's setup hook.

2. Cyclic dependencies are now an error.  Yes, they are only possible in absurd cases, but possible nonetheless:
   ```
   └── /nix/store/asdfasdfasdf-foobar/
       ├── foo
       │   └── nix-support
       │       └── propagated-build-inputs = /nix/store/asdfasdfasdf-foobar/bar
       └── bar
           └── nix-support
               └── propagated-build-inputs = /nix/store/asdfasdfasdf-foobar/foo
   ```
   (`foo` and `bar` are cyclic). The old algo would silently work with this; the new one would naively diverge, so I made it error out instead as such a thing should never be intended. But if we agree this shouldn't happen in practice so infinite recursion is a good enough error, I can simplify things

###### Things done

~~Hoping to merge this **Monday 2017-11-13** at the latest :). Thanks!~~

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built stdenv on platform(s)
   - [ ] NixOS -- latest in progress, see nixborg build
   - [ ] macOS -- latest in progress
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---------

CC @bgamari

Ericson2314 committed:
bintools-wrapper, cc-wrapper: Use propagatedBuildInputs to simplify setup hook

This will affect the path, and anything else should be harmless. I do
wonder why this is needed at all with the wrapper scripts doing their
own PATHs however.

Ericson2314 committed:
stdenv setup: Ensure each pkg's deps are processed first

This would be changing from a "preorder traversal" to a "postorder
traversal".

We need to *prepend* to `_PATH` now so that shallower entries take
priority. Other appends that should be prepends will be fixed in the
following commits.

Ericson2314 committed:
stdenv setup: Guard against dependency cycles

As of the last stdenv setup.hs commit, the package accumulation arrays
now only contain *fully* processed packages, but if a package is a
dependency of itself, it is only *partially* progressed when the
function encounters it again. Normally, dependencies are entire
derivations, ensuring the dependency graph is in fact acyclic, but it's
perfectly possible to stuff two dependencies in the same derivation
output such that each depends on each other.

Maybe that's an absurd thing no one would do, but I wanted to be
conservative introducing this change so I added a separate stack for
partially visited packages, ensuring that we'd error out on encountering
such a thing. If we actually find this is done intentionally, it would
be easy to change behavior such that the package silently isn't
traversed again.

copumpkin commented:
This seems sensible in principle, but can we get a Hydra job showing what the implications might be? Subtly changing behavior (that arguably nobody should rely on the order of, but people probably do) makes me very uncomfortable even if it does make sense.

globin commented:
@nixborg build

globin commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

orivej commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -310,20 +313,54 @@ findInputs() {
+}
+
+# Add package to the future PATH and run setup hooks
+activatePackage() {
```
Who calls this function?

Ericson2314 commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -359,15 +389,22 @@ if [ -z "${crossConfig:-}" ]; then
         findInputs "$i" nativePkgs propagated-native-build-inputs
     done
+    for i in "${nativePkgs[@]}"; do
+        activatePackage "$i"
```
@orivej here

Ericson2314 commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -359,15 +389,22 @@ if [ -z "${crossConfig:-}" ]; then
         findInputs "$i" nativePkgs propagated-native-build-inputs
     done
+    for i in "${nativePkgs[@]}"; do
+        activatePackage "$i"
```
@orivej And here

Ericson2314 commented:
@globin is there any good known-good jobsest to compare the nixborg one to?

Ericson2314 commented:
Oh hmm, this breaks down when different dependencies are in fact the same derivation. I'm going to have to think about that.

orivej commented:
Could you give an example in nixpkgs?

Ericson2314 commented:
@orivej Nevermind. The already-visited logic *is* based on actual paths, I just mistakenly changed the iteration order. [I did have to make `_PATH` be prepended instead of appended, but that is expected.]

There's still a theoretical issue where "unrelated" deps could clobber each other on the PATH in different ways based on different topological sorts, and our depth-first one might not have the most intuitive behavior in that regard, but I won't worry about this unless it shows up in practice.

I added a `cc-wrapper` commit from my https://github.com/NixOS/nixpkgs/pull/29396 branch that actually leverages this, to make sure everything works alright.

orivej reviewed (approved):
Looks good. https://hydra.mayflower.de/jobset/nixos/pr-31414 did not pick your changes yet… And since AFAIK it does not contribute to the nixos cache, shouldn't this PR go to staging first?

globin commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

Ericson2314 committed:
treewide: Use `prependToSearchPath*` instead of `appendToSearchPath*` in env hooks

Now that the packages are accumulated with respect to dependency order,
later packages should supercede older packages.

Ericson2314 committed:
treewide: Rename `addToSearchPath*` to `appendToSearchPath`

The following commit will add a way to prepend, so this makes the
choice explicit.

Ericson2314 commented:
@nixborg build

Ericson2314 commented:
Ok, the last build failed because of uses of `addToSearchPathWithDelimiter`. I renamed `addToSearchPath*` to `appendToSearchPath` across the board to be explicit and consistent, avoiding the undefined function problem in the process. This does break the the interface, but as those functions aren't really documented anywhere I hope there was no expectation of stability.

Additionally, I changed some `appends*` to `prepends*` to match the new near-opposite order dependencies are accumulated as. I think there's a good chance that fixes the only potential breakage this PR introduces. [The Nixborg packages were not affected by this either way as the first hydra evaluation shows, oh well.]

Still hoping to merge this tomorrow, but I'll understand if this change to the PR means people need more time to review. I have reorganized commits so the history still tells a clear story.

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

Ericson2314 commented:
Hmm the darwin error is seems to be https://github.com/Andersbakken/rtags/issues/937#issuecomment-293459813, a problem with `#include_next` and the incnlude order.  Still investigating the Linux problem.

My last attempt did not work as well as I thought...

Ericson2314 committed:
treewide: prepend, not append PATH-like variables in setup hooks

The previous commit replaced `appendToSearchPath*` calls to this effect.
This one manually fixes up remaining more manual appends, including
using `prependToSearchPath*` instead of manual logic where possible.

Ericson2314 commented:
Ok, I I did two things which should fix the previous problems:

1. Default dependencies are processed first, so they are still lower priority. This should fix the Darwin problem as it will effectively put libc before just about any other library.

2. Setup hooks are (painstakingly!) edited to ensure the current package supersedes any previously-processed one.

Ericson2314 commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

Ericson2314 commented:
@nixborg build

Ericson2314 commented:
Rebased on top of now-merged https://github.com/NixOS/nixpkgs/pull/31723, which was part of this. I doubt that will fix this, but if it does, great, and if it doesn't, the build will fail quickly so that not too many resources were wasted.

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

FRidh commented:
I like the motivation behind this change but because nothing is happening I am closing this.

JohnAZoidberg commented:
This issue is mentioned in nixpkgs stdenv: https://github.com/nixos/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L1167

Ericson2314 commented:
Yes maybe a GitHub issue rather than open PR canbe made for that.

Ericson2314 commented:
Reopning and marking as draft because I ought to finish it at some point.

Ericson2314 commented:
Should be possible to finish after #132340 is the default and cc-wrapper is simplified.

ConnorBaker commented:
I think this is still relevant and (time permitting) I'd like to revive this PR.

@Ericson2314 in terms of how this would look on current `master`, any advice? I started by trying to apply the spirit of the changes each commit contains, but I'm not sure what some of the changes should look like post #26805.

For example, what should https://github.com/NixOS/nixpkgs/pull/31414/commits/229eed32411b2d680757df437073a818a080c0ce look like since https://github.com/NixOS/nixpkgs/pull/26805/commits/3a50395ef226259694689da92b2fd636476abc67? That is, I'm not sure what the offsets are for the wrapper and scripts relative to the consumers.

Additionally, I see a lot of interpolation of derivations into store paths in various strings -- is this a concern with respect to splicing? I know that `mkDerivation` includes code to handle retrieving packages from the correct package set when they are in dependency arrays, but I don't think that applies to the various string arguments passed to `mkDerivation` (nor could it, I suppose).

Ericson2314 commented:
> That is, I'm not sure what the offsets are for the wrapper and scripts relative to the consumers.

I think the offsets can stay the same? Or at least if

- the old PR moved things to propagated deps
- `master` had recategorized which type of dep

Do "propagated" per the PR, and the offset per `master`.
